[MM-13507] Plugin framework: auto-restart plugins that crash (original PR)

mickmister opened the pull request:
#### Summary
This PR adds new functionality to monitor running plugins, and be able to take action when they are not responding. With this new functionality, all plugins are periodically monitored, and restarted or deactivated based on their health status. 

If a plugin is deactivated, it is shown with an error in the Admin Console. The admin can then restart the plugin if it is a situation they will solve on their end.

#### Ticket Link
[Github Ticket](https://github.com/mattermost/mattermost-server/issues/10386)
[Jira Ticket](https://mattermost.atlassian.net/browse/MM-13507)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

#### This PR monitors the plugin’s health in 3 different ways:
- Using the plugin process's pid to test if the process will respond to a zero signal (may have different behavior per OS)
- Using go-plugin's Ping method to test RPC connection
https://github.com/hashicorp/go-plugin/blob/master/rpc_client.go#L167
- Using a new HealthCheck hook if the plugin implements it

mickmister commented:
#### Example when plugin process is not responding:
<img width="798" alt="Screen Shot 2019-03-17 at 5 58 36 PM" src="https://user-images.githubusercontent.com/6913320/54499258-07ad8100-48e6-11e9-83b6-7e999e359536.png">

#### Example when plugin HealthCheck hook is returning a specific error, like a configuration error:
![image](https://user-images.githubusercontent.com/6913320/54499269-2e6bb780-48e6-11e9-8271-6eb67ae7beea.png)

#### Example when plugin is healthy:
![image](https://user-images.githubusercontent.com/6913320/54499279-44797800-48e6-11e9-8733-564eaf8ca880.png)

mickmister commented on plugin/environment.go:
```diff
@@ -247,6 +256,52 @@ func (env *Environment) Deactivate(id string) bool {
+}
+
+// Checks for plugin error status by using the two above functions, as well as the HealthCheck hook, if implemented.
+func (env *Environment) GetPluginErrorStatus(id string) error {
```
Here is where the different kinds of health checks are made. The flow currently goes:

```
if err1 = RPC ping fails:
  if err2 = plugin process not responding:
    return err2
  else
    return err1

if err3 = HealthCheck hook returns error
  return err3
```

mickmister commented on app/health_check.go:
```diff
@@ -0,0 +1,92 @@
+	taskMutex sync.Mutex
+}
+
+const (
```
These variables will need to be extracted into a configurable policy.

mickmister commented on model/bundle_info.go:
```diff
@@ -11,6 +11,8 @@ type BundleInfo struct {
 	ManifestPath  string
 	ManifestError error
+
+	HealthCheckFails int
```
Not stateless here. I believe we need to store the health check failure elsewhere so it will work in a distributed context.

crspeller replied:
Certainly the bundle info should be read-only.

mickmister replied:
Right, this is a placeholder per plugin until it is decided where to store this info. It definitely doesn't feel right to store this here.

jwilander replied:
Yeah I don't think we need to worry about a distributed context here, the health check is for the plugin running on a specific app server so only that app server cares

mickmister commented on plugin/supervisor.go:
```diff
@@ -17,9 +19,11 @@ import (
-	hooks       Hooks
-	implemented [TotalHooksId]bool
+	client         *plugin.Client
+	clientProtocol plugin.ClientProtocol
```
This is used to access the `Ping()` method implemented here:
https://github.com/hashicorp/go-plugin/blob/master/rpc_client.go#L167

The `ClientProtocol` interface is used to be ambiguous to whether the client is a `RPCClient` or `GRPCClient`. There is an [issue in mattermost-server](https://github.com/mattermost/mattermost-server/issues/9486) made to support gRPC. The `ClientProtocol` interface used here may be relevant for going forward with support for gRPC.

mickmister commented:
Copying over a comment from the ticket made by @harshilsharma63:

> Sometimes only the server component of the plugin crashes while the webapp part keeps running. Would we restart the plugin if either of them crashed?

crspeller commented on app/health_check.go:
```diff
@@ -0,0 +1,92 @@
+
+func (job *PluginHealthCheckJob) Start() {
+	mlog.Debug(fmt.Sprintf("Plugin health check job starting. Sending health check pings every %v seconds.", PLUGIN_HEALTH_CHECK_INTERVAL))
+	newTask := model.CreateRecurringTask(PLUGIN_HEALTH_CHECK_TASK_NAME, job.RunPluginHealthChecks, time.Duration(PLUGIN_HEALTH_CHECK_INTERVAL)*time.Second)
```
Not sure the job system is the right fit here. It's designed to run once on an entire cluster and possibly on a separate server. But this needs to run on each server.

mickmister replied:
That makes sense, thanks for pointing this out. Do you have ideas on how to schedule the health check per server?

crspeller replied:
I don't think we have a specific pattern for that. Just a regular go routine would work.

mickmister replied:
@levb I agree. I'm thinking of how to solve this. It would likely be a function of the health check interval and the number of plugins. Or possibly an explicitly configured value for the time distance between each plugin's check. There is also the situation where we check upon visiting the admin console. In this case, they would be simulataneous yet concurrent. In the health check job situation, they would also be concurrent, if simulataneous.

crspeller commented on plugin/supervisor.go:
```diff
@@ -17,9 +19,11 @@ import (
-	hooks       Hooks
-	implemented [TotalHooksId]bool
+	client         *plugin.Client
+	clientProtocol plugin.ClientProtocol
```
Why store this and not call `client.Client()`?

mickmister replied:
> Why store this and not call client.Client()?

Good call. I had missed the portion of the `Client()` [method](https://github.com/hashicorp/go-plugin/blob/3f118e8ee104b6f22aeb12453fab56aed1356186/client.go#L349) that short circuits if the `ClientProtocol` is already assigned.

levb commented on plugin/supervisor.go:
```diff
@@ -103,6 +112,24 @@ func (sup *supervisor) Hooks() Hooks {
 }
 
+// Pings the plugin through RPC
+func (sup *supervisor) Ping() error {
```
What information or other benefit do we get from Ping that we do not get from invoking the Healthcheck hook?

mickmister replied:
Most plugins will not have the HealthCheck hook implemented, especially at the time of initial deploy of this PR's functionality. If we can make a default implementation that does some diagnostics, or some secondary HealthCheck hook that is always our own implementation, then that can take precedence. Though, I believe the `Ping` method is useful because:
1. The method is guaranteed to be available without any further code on our part.
2. It specifically tests that the RPC connection is still alive and healthy. That is its sole purpose in the go-plugin library.
3. It's quick and cheap, other than network or timeout issues.

hanzei commented on app/health_check.go:
```diff
@@ -0,0 +1,92 @@
+	ENABLE_HEALTH_CHECK_JOB       = false
+)
+
+func (s *Server) InitPluginHealthCheckJob() {
```
I was wondering the same. What do you think about adding this https://github.com/mattermost/mattermost-server/blob/master/plugin/supervisor.go#L19-L23

hanzei commented on app/health_check.go:
```diff
@@ -0,0 +1,92 @@
+
+		if plugin.HealthCheckFails == HEALTH_CHECK_FAIL_LIMIT {
+			mlog.Debug(fmt.Sprintf("Plugin `%v` exceeded health check failure threshold. Restarting Plugin.", id))
+			pluginsEnvironment.RestartPlugin(id)
```
If I understand the logic correct, this will ensure the plugin get restarted every time it has failed 3 time? Do we really want to restart it _forever_? Maybe just add a restart limit instead?

levb replied:
THAT ^^. There needs to be a policy for retrying, and a policy for restarting, both configurable. The loop of restart/crash/detect must be avoided.

mickmister replied:
@levb I agree. This should be a main focus of this monitoring/restart effort. Once that loop has begun, it will happen forever as @hanzei pointed out. As such, I'm treating this specific issue with very high importance for design moving forward.

hanzei replied:
My thoughts on the policy are:

For me the three cases (Ping failed, health check failed, process not found) all require separate policies.

1. Ping failed: Retry ping A times, then restart
2. Health check: Restart immediately but only B times in C minutes. If the limit B is reached, disable the plugin.
2. Health check: Same as 2. but with a different limit.

Thoughts on this?

mickmister replied:
I have the different failure cases implemented with a common shared fail limit at the moment, but I think they could benefit by treating them as separate policies as you mention. The reasons for failure are different, and shouldn't necessarily be handled the same way. I'd be curious to hear others' thoughts on this.

hanzei commented on app/plugin.go:
```diff
@@ -295,6 +295,17 @@ func (a *App) DisablePlugin(id string) *model.AppError {
 
+// RestartPlugin will deactivate and reactivate a plugin.
+func (a *App) RestartPlugin(id string) *model.AppError {
+	pluginsEnvironment := a.GetPluginsEnvironment()
```
It fails if `!*a.Config().PluginSettings.Enable`.

What do you propose as NS: Create a ticket? Solve this in this PR?

hanzei commented on app/health_check.go:
```diff
@@ -0,0 +1,92 @@
+			return
+		}
+
+		err := pluginsEnvironment.GetPluginErrorStatus(id)
```
Maybe log the error? At least as debug.

hanzei commented on plugin/hooks.go:
```diff
@@ -139,6 +140,11 @@ type Hooks interface {
+	// HealthCheck is invoked periodically to assert the plugin's current health.
+	// If the health check fails more than the configured number of times, the process will be destroyed and restarted.
+	//
+	HealthCheck() error
```
I can't think of a use case where restarting the plugin helps with the problem. Can someone please name one for me?

levb replied:
The plugin maybe deadlocked, process running but not responsive.

jwilander replied:
Could also be a rare crash where the issue only happens infrequently so restarting the plugin brings it back up for common case use

hanzei replied:
> The plugin maybe deadlocked, process running but not responsive.

If the plugin can detect it's own deadlock, why shoun't it just `panic`?

My questions is: Why should we poll periodically if the plugin is unhealthy, instead of waiting for the plugin to send us this information? (Via e.g. a `Exit()` API method)

hanzei replied:
Sorry for being persistent here, but I would like to remove this discussion before continuing with further review.

@levb @jwilander @mickmister I would love to hear you thoughts on this.

mickmister replied:
If we were to support the Exit() API method, can we assert that it is being used, much less being used properly on the plugin side? I think adding more methods to be able to gather info on the current state of the plugin is right, but I don't think we should throw out the mechanism of monitoring it directly, from the outside.

I think the process check and regular RPC ping are the most useful, because it will work out of the box for all plugins. To me, the HealthCheck hook (and the Exit() API  method) are additional pieces that require the plugin author to follow an optional convention. That may change but it should probably remain backwards compatible.

Re-reading the context of your question, you may just be talking about the HealthCheck hook and not the other two mechanisms. If the plugin is written in a way that can tell us when it is crashing, that makes sense. But if it is unhealthy by some other metric or it is not panicking/crashing, asking it periodically seems to be a good way to find out about those issues, so it doesn't have to run its own interval to continuously run those checks.

mickmister replied:
Another option for the plugin author would be to have the plugin run its own checks at the beginning and/or end of each hook/http call. Then the plugin could call `Exit()` whenever it saw fit, after a self-check determined it was unhealthy. Is this what you had in mind @hanzei?

I like this approach because the plugin author or admin will know the context of where in execution it is exiting, versus arbitrarily running a check when the server has asked it to. Both approaches have their own merits, and they certainly aren't mutually exclusive.

The HealthCheck hook has the potential to collect info on the plugin's status, without it necessarily being an exit condition. The hook would need to support a return type that signals that it does not need to be restarted. This could, in the future, integrate with external monitoring systems to support thorough diagnostic reporting for admins, including health of the Mattermost server itself.

hanzei replied:
> Then the plugin could call `Exit()` whenever it saw fit, after a self-check determined it was unhealthy. Is this what you had in mind @hanzei?

As hooks are the only "entry point" for executing plugin code: Yes :wink:.

> The HealthCheck hook has the potential to collect info on the plugin's status, without it necessarily being an exit condition.

Maybe I misunderstood your proposal, but I thought the plugin would be restarted every time the `HealthCheck` returns an error. Or are you talking about just collecting data and not returning an error?

> The hook would need to support a return type that signals that it does not need to be restarted. 

I really like this idea. :+1: This would also close the discussion about when to restart the plugin and when not.


Do we agree that a `Exit` API method is a good idea? Maybe with the same return type that signals that it does not need to be restarted?

If so, do you want to include this in this PR or just open a ticket for this?

mickmister replied:
I like the idea of using a push protocol to communicate to the Mattermost server that there is an issue, so I'm all for the `Exit()` method. If this method would support a non-restart field, then I suggest we name it something other than `Exit`. It's making me think that we should support three situations here:

- trivial error, no action necessary other than reporting the error
- fatal error, restart plugin
- fatal error, kill plugin without restarting

In any case, these messages should be reported to the admin somehow, which will be covered in a follow-up PR addressing UX issues around these monitoring/restart changes. I would say that supporting this API method in general is out of scope of this PR.

@levb, any thoughts on this?

levb replied:
I really like the 3 conditions described, 2/5 I'd return an `error` and a separate `bool` to `DoNotRestart` (0/5 on the default behavior and therefore the naming).

Not sure I am a fan of push notifications and/or plugins managing their own restarts, in fact 1/5 predisposed against both. IMO ideally, like any production code, plugins should be free of panics. IMO pull is generally a more predictable way of reliably gathering data for centralized processing. 0/5 pull results in less code.

hanzei replied:
> I really like the 3 conditions described, 2/5 I'd return an `error` and a separate `bool` to `DoNotRestart` (0/5 on the default behavior and therefore the naming).

But we need more then a `bool` to describe the three conditions. Or do I get your proposal wrong?

> IMO ideally, like any production code, plugins should be free of panics. 

I don't see them as panics. Think about e.g. a plugin requiring a specific config setting. I would check in `OnConfigurationChange` if this setting is pressed and otherwise `Exit()'. 

> IMO pull is generally a more predictable way of reliably gathering data for centralized processing.

Could you elaborate a bit?

mickmister replied:
> But we need more then a bool to describe the three conditions. Or do I get your proposal wrong?

I agree, a single boolean would not allow for all cases to be handled. I think it's better to think of it as four conditions, where one condition is that no error occurred. So it would be something like this:
- no error: `return nil, nil`
- trivial error: `return err, nil` or `return err, Actions.LEAVE_RUNNING`
- error that warrants restart: `return err, Actions.KILL_AND_RESTART`
- error that should just kill plugin (like config error) `return err, Actions.KILL_WITHOUT_RESTART`

However, this implementation seems a bit awkward. We may not want the plugin author to be able to differentiate between `kill` and `kill and restart`, if we  want to avoid this level of granularity.

hanzei replied:
> We may not want the plugin author to be able to differentiate between `kill and restart` and `kill`.

Why would we not want allow plugins to differ between these two? E.g. if a plugins deadlock it would call `kill and restart` and if I see a not supported config value it would call `kill'.

mickmister replied:
@hanzei I agree, please disregard my comment about limiting the options. We would want to give options to solve those cases as you point out.

To account for all three error cases, I'm thinking we can use an enum, similar to the [constants](https://github.com/mattermost/mattermost-server/blob/387482f77b2239182100cb99252d390601fb1bd3/plugin/hooks.go#L17) that are used as ids to differentiate between the available hooks. An example for signaling a specific error case would be something like `PLUGIN_HEALTH_CHECK_IGNORE_ERROR`, `PLUGIN_HEALTH_CHECK_KILL`, and `PLUGIN_HEALTH_CHECK_RESTART`. What do you think of this method?

hanzei replied:
:+1: for enums. I would like to use a typed one to make clear what the expected values are. Like described in https://stackoverflow.com/a/14426447. Are you fine with this?

To be fair: I have no good idea for naming. I would like to avoid `HEALTH_CHECK` as I want to use the same enum for an `Exit()` method later. Maybe something like `PLuginExitCommand`? 0/5

hanzei replied:
One additional thought:

Why do we need a specific enum anyway? Why not go with two methods `Exit()` and `Restart()`. The error returned in the healthcheck hook would just be logged. And if a plugin needs to exit in some way, it can call the API methods.

I know this idea is a bit late in the discussion, but this design would make it much more clear, what the healtheck does and what not.

mickmister replied:
At the point of writing, it's been established that a failed process check or RPC ping should warrant a plugin restart, but there is discussion in this thread on other cases where the plugin should be restarted. The two additional methods being discussed are:

- allow restarting of the plugin by the server, after retrieving an error from the HealthCheck hook (after process check and ping pass)
- allow the plugin author to choose when to restart or kill based on new plugin API methods `Exit()` and `Restart()` and the HealthCheck hook would not restart the plugin

Along with the idea of allowing the HealthCheck hook result to restart the plugin, there is also discussion on letting the returned error have a conditional signal to tell the server what action to take on an unhealthy plugin. The server could use the signal to kill and restart the plugin, use the signal to explicitly kill the plugin without restarting, or use the signal to say that the error is benign and shouldn't be used to kill or restart the plugin. Using API methods from the plugin side to restart instead of the HealthCheck hook would mean that we don't need these special flags.

I'm inclined to remove the HealthCheck hook from this PR and keep it as just the process check and ping for the purposes of purely introducing code for monitoring and restarting plugins. Adding any application-level options increases the scope of discussion.

@crspeller @jwilander Do you have thoughts on these points, or thoughts on what should be considered in the scope of this PR in regards to this discussion?

hanzei replied:
> I'm inclined to remove the HealthCheck hook from this PR and keep it as just the process check and ping for the purposes of purely introducing code for monitoring and restarting plugins. Adding any application-level options increases the scope of discussion.

100% agree on this. Sorry for driving away the discussion from the original idea of this PR.

jwilander commented on app/health_check.go:
```diff
@@ -0,0 +1,92 @@
+)
+
+func (s *Server) InitPluginHealthCheckJob() {
+	if ENABLE_HEALTH_CHECK_JOB {
```
I imagine this constant is just part of the WIP bits and this would instead be a configurable setting?

mickmister replied:
Right. This was set here as an ease for initial implementation. This push was mainly meant to gather more in-depth requirements and generate discussion within the team.

As for this specific suggestion, are you thinking that the configuration for fields such as backoff delay, health check interval, and restart limit should be configured per plugin? If it is per plugin, the configuration could be stored in the specific plugin's `plugin.json` file, with defaults stored in the server's `config.json` file. For the "global" config of these health check settings, it could be stored in the server's `config.json` like so:

```
{
  "PluginSettings": {
    "HealthCheckSettings": {
      "HealthCheckInterval": 30,
      "HealthCheckFailLimit": 3 (how many failed checks before we should restart, excluding short-circuiting fatal checks like process does not exist),
      "HealthCheckRestartLimit": 3 (how many restarts before we say that this plugin will continue to fail),
      "HealthCheckRestartWaitTime": 60 (how long we wait for plugin to be considered healthy before we decide to restart again, because we don't want to restart again if it fails on startup),
      "EnableHealthCheckJob": true,
      "EnableHealthChecksInAdminConsole": true
    }
  }
}

```
There are definitely more configuration options we could include, and some listed above may not be included or have overlapping use cases with other config options.

jwilander replied:
Actually ideally I think I'd rather not expose it as configuration settings to the admin or plugin authors at all and just take care of it ourselves. I don't think admins or plugin authors would better be able to set these than us, so I think this is a spot where we should be opinionated and just pick a set strategy

jwilander commented on plugin/environment.go:
```diff
@@ -247,6 +256,52 @@ func (env *Environment) Deactivate(id string) bool {
+	return fmt.Errorf("plugin not running: %v", id)
+}
+
+// Checks for plugin error status by using the two above functions, as well as the HealthCheck hook, if implemented.
```
To me if the check for the process fails that's a failure state right there, no need to wait for multiple health checks to confirm

mickmister replied:
I agree. If the process is dead, there's no point in doing anything other than some recovery or clean up action.

mickmister commented:
Hi @jwilander @levb @crspeller @hanzei,

I've added some updates to address suggestions for the restart policy. I've also tacked on some lingering questions to the end of the PR description.

hanzei commented:
Will take a look next week.

mickmister commented on app/server.go:
```diff
@@ -188,6 +188,18 @@ func NewServer(options ...Option) (*Server, error) {
+		if pluginsEnvironment != nil {
+			pluginsEnvironment.InitPluginHealthCheckJob(c.PluginSettings.EnableHealthCheck)
+		}
+	})
```
Will running this in server.NewServer() run this on every instance in a cluster? The EmailBatching job begins this way, which is what I initially modeled the job after.

crspeller replied:
Yes. This should now run on every instance.

crspeller commented on plugin/plugin_health_check.go:
```diff
@@ -0,0 +1,182 @@
+)
+
+const (
+	PLUGIN_HEALTH_CHECK_INTERVAL  = 5 // minutes
```
Seems a bit long. I would want to know if my plugin was unhealthy faster than that. Maybe 1 minute? 30 seconds?

mickmister replied:
👍 30 seconds sounds good to me

crspeller commented on plugin/plugin_health_check.go:
```diff
@@ -0,0 +1,182 @@
+		env := job.env
+		activePlugins := env.Active()
+
+		for _, plugin := range activePlugins {
```
Maybe it would be more clear that this is one time code if it was outside the `go func()`.

mickmister commented on plugin/environment.go:
```diff
@@ -113,13 +115,22 @@ func (env *Environment) Statuses() (model.PluginStatuses, error) {
+			if err := env.CheckPluginHealthStatus(plugin.Manifest.Id); err != nil {
+				pluginErr = err.Error()
+				pluginState = model.PluginStateError
+			}
```
I'm thinking this may need to be i18n'd since these errors would be returning to the UI. The kind of error that would be here is something like "Plugin process not found, or not responding", or something returned from the HealthCheck hook.

This PR has been mainly focused on the auto-restart capabilities of the plugins, and seems to have shifted some focus away from the UX from the admin console of managing the failed plugins. Any thoughts on making the `api/v4/plugins/<plugin_id>/restart` api route, and the UX around including error info in the `api/v4/plugins/<plugin_id>/statuses` route into a separate PR to keep the scope of this PR more isolated? However, that PR would depend on this PR's functionality.

mickmister replied:
There's some more plugin framework work to be done here as well, like docs and testing of the HealthCheck hook, and possibly adding the `RestartPlugin` method to the PluginAPI.

crspeller replied:
Ya. If you want to do the UI as a separate PR so this can get merged. That would work.

mickmister replied:
@crspeller Would you suggest I remove the restart api route from this PR, or keep it how it is and continue development on it when this is merged?

mickmister commented on plugin/supervisor.go:
```diff
@@ -103,6 +110,63 @@ func (sup *supervisor) Hooks() Hooks {
+			err = sup.Ping()
+			if err == nil {
+				break
+			}
```
I'm thinking we should space out the pings a bit and account for timeouts. Something like this:

```go
var pingErr error
for pingFails := 0; pingFails < HEALTH_CHECK_PING_FAIL_LIMIT; pingFails++ {
  c := make(chan error, 1)
  go func() {
    c <- sup.Ping() 
  }()

  select {
  case err := <-c: // Ping() either successful, or explicitly failed before timeout
      if err != nil { // Ping() explicitly failed
          pingErr = err
          time.Sleep(50 * time.Millisecond) // sleep for 50ms
      } else { // Ping() was successful
          break
      }
  case <-time.After(50 * time.Millisecond): // account for 50ms timeout
      pingErr = errors.New("Ping timed out")
  }
}

if pingErr != nil {
  mlog.Debug(fmt.Sprintf("Error pinging plugin, error: %s", pingErr.Error()))
  return fmt.Errorf("Plugin RPC connection is not responding")  
}
```

crspeller replied:
Doesn't Ping timeout itself?

mickmister replied:
Looking at the source for go-plugin and net/rpc, we should be good to not include the timeout check.

How do you feel about including the spaced out ping calls here with `time.Sleep()`?

mickmister replied:
I'm thinking that the timeout isn't appropriate here for Ping, but with the case of the HealthCheck hook, the plugin author may have an issue that causes a never ending block during the hook's execution. We would want to avoid that as I believe it would cause the health check loop to block indefinitely.

mickmister commented on plugin/plugin_health_check.go:
```diff
@@ -0,0 +1,182 @@
+
+	supervisor := ap.supervisor
+
+	pluginErr := supervisor.PerformHealthCheck()
```
Similar timeout check for the overall health check:

```go
var pluginErr error
c := make(chan error, 1)
go func() {
    c <- supervisor.PerformHealthCheck()
}()

select {
case err := <-c:
    pluginErr = err
case <-time.After(250 * time.Millisecond):
    pluginErr = errors.New("Health check timed out")
}

```

mickmister commented on plugin/hooks.go:
```diff
@@ -139,6 +140,11 @@ type Hooks interface {
+	// HealthCheck is invoked periodically to assert the plugin's current health.
+	// If the health check fails more than the configured number of times, the process will be destroyed and restarted.
+	//
+	HealthCheck() error
```
Proposed rewording for docs:

> HealthCheck is invoked every 30 seconds to assert the plugin's current health.
If the health check fails, the process will be destroyed and restarted.
If the health check fails more than three times within one hour, the plugin will be deactivated until an admin chooses to restart it.


We could choose to be less specific with describing the criteria of deactivating the plugin. Though, I'm thinking that it would be good to know exactly how the health check interval and restart policy works from the perspective of a plugin author. Once there is a consensus on the values chosen, then the wording can be finalized.

crspeller commented:
@mickmister 
I think the conversation is getting a bit lost in all the specific feedback. Do you mind summarizing what the outstanding issues and questions are so we can get them addressed and get this merged?

mickmister commented:
Thank you for suggesting this @crspeller 

Here are some outstanding issues/questions:

1. Waiting between pings: Currently, when the plugin fails a single ping test, the next ping is sent right away without any wait time. Do we think we should put a small wait time, such as 10-20ms, between each ping? This would give the RPC connection a moment to recuperate if it is experiencing heavy traffic.

2. Handling timeouts of the HealthCheck hook: When running the HealthCheck hook, there may be an issue on the plugin implementation’s side that causes the hook to never return/finish, creating a block in the health check loop. Should we account for the HealthCheck hook indefinitely blocking by using a timeout around the RPC call? I’m not sure if this issue is common when running hooks, so I’m not sure how practical this is. I believe @levb had touched on handling timeouts a bit in one of our previous discussions.

3. Currently, the health check interval is set up to run every 30 seconds. If the check fails 3 times in 1 hour then the plugin will finally be deactivated without a restart. I’d like to get some thoughts on these numbers chosen, namely the 1 hour choice.

4. There is some discussion on having the HealthCheck hook’s return type support more than just errors. Do we think this change should be included in this PR? This functionality will need some more design around it, so I’m thinking this should be pushed out to simplify the review of this PR.

5. There is a follow-up PR coming after this pertaining to the admin's UX around reviewing and resolving the errors found in the health checks. There is some related functionality in this PR, namely introducing the `plugins/<plugin_id>/restart` api route, and introducing health check error info in the `plugins/<plugin_id>/statuses` route. I’m wondering if these pieces should be outright removed from this PR, or should it be left as is and continued after this is merged? For the sake of the scope of the PR I don’t think it makes sense for it to be in the PR, but I think I need a second opinion on whether it matters or if it should be included here.

levb commented:
Will respond tomorrow

mickmister commented:
I was getting into some primitive unit testing for this yesterday. It's a lot more feasible to write realistic tests than I originally thought, without the need of placing interfaces everywhere. I'm going to dive into it more tonight and hopefully have some tests for review by the beginning of next week.

mickmister commented:
Here are some updates on the issues laid out above:

1 and 2:
I say let's not include these in this PR, unless we want to change how the health check works in general. @levb and I had a discussion about how the job may work if we were to get rid of the interval and have a goroutine per plugin-based approach. This would change the code quite a bit, since it currently works as a sequential synchronous check of all of the plugins. I'm open for discussion on any suggestions here.

3:
Please see the review comment below to discuss the constants used in the health check job.

4:
Having the HealthCheck hook return special-case errors seems to be out of scope to me. I say we shouldn't include this functionality here. @hanzei What are your thoughts on this?

5:
I have removed the code pertaining to displaying plugin errors and restarting from the Admin Console, as it is out of scope for the purpose of this PR.

I've removed WIP from the title as I believe this PR is ready for review.

mickmister commented on plugin/health_check.go:
```diff
@@ -0,0 +1,165 @@
+	PLUGIN_HEALTH_CHECK_INTERVAL  = 30 // seconds
+	HEALTH_CHECK_PING_FAIL_LIMIT  = 3
+	HEALTH_CHECK_RESTART_LIMIT    = 3
+	HEALTH_CHECK_DISABLE_DURATION = 60 // minutes
```
Here's a description for each variable here:

PLUGIN_HEALTH_CHECK_INTERVAL - How often we start the sequential check of all plugins

HEALTH_CHECK_PING_FAIL_LIMIT - How many failed calls to `client.Ping()` in a given plugin's health check before we deem the RPC connection to be unhealthy

HEALTH_CHECK_RESTART_LIMIT - How many times we restart the plugin before permanently deactivating it

HEALTH_CHECK_DISABLE_DURATION - How long we allow successive failures to occur before we permanently deactivate the plugin

hanzei commented:
4. Fine by me to leave this out. My only concern is, that once we add this in, we kind of remove all old restart logic of the hook.

mickmister commented:
I've decided to create a new PR for a more defined scope of the task, based on discussion. I figure it is better to have a clean commit to review its current state, and any further changes will be much easier to review. https://github.com/mattermost/mattermost-server/pull/10781

@crspeller @hanzei @levb
